File: 1001-1100/1061-Lexicographically Smallest Equivalent String/1061-Lexicographically Smallest Equivalent String.cpp
class Solution {
public:
    string smallestEquivalentString(string A, string B, string S) {
        vector<int> parent(26);
        for (int i = 0; i < 26; ++i) {
            parent[i] = i;
        }
        
        for (int i = 0; i < A.size(); ++i) {
            int p1 = findParent(parent, A[i] - 'a');
            int p2 = findParent(parent, B[i] - 'a');
            if (p1 < p2) {
                parent[p2] = p1;
            }
            else if (p1 > p2) {
                parent[p1] = p2;
            }
        }
        
        string result;
        for (char c : S) {
            result += 'a' + findParent(parent, c - 'a');
        }
        return result;
    }
    
private:
    int findParent(vector<int>& parent, int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
};

File: 1001-1100/1061-Lexicographically Smallest Equivalent String/1061-Lexicographically Smallest Equivalent String.py
import string
class Solution:
    def smallestEquivalentString(self, A: str, B: str, S: str) -> str:
        parent = {c: c for c in string.ascii_lowercase}
        def findParent(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        
        for c1, c2 in zip(A, B):
            p1 = findParent(c1)
            p2 = findParent(c2)
            if p1 < p2:
                parent[p2] = p1
            elif p1 > p2:
                parent[p1] = p2
                
        return ''.join(findParent(c) for c in S)
